src/main/java/de/theit/jenkins/crowd/CrowdServletFilter.java

```diff
@@ -178,6 +178,15 @@ public void doFilter(ServletRequest request, ServletResponse response,
                         if (null != auth) {
                             LOG.log(Level.FINE, "User successfully logged in");
                             sc.setAuthentication(auth);
+                        } else { // Auto login failed.
+                            if (LOG.isLoggable(Level.FINE)) {
+                                LOG.fine("User failed to log in");
+                            }
+                            HttpSession session = req.getSession(false);
+                            if (session != null) {
+                                session.invalidate();
+                            }
+                            SecurityContextHolder.clearContext();
```

Comment:
the only thing i'm missing here is cache clearing
If you invalidate user it still have records in cache which together with token caching can bypass sso

Reply:
> the only thing i'm missing here is cache clearing
I'm getting you the cache clearing.
Do you mean the crowd2-plugin caches tokens somewhere else other than in the session?

> If you invalidate user it still have records in cache which together with token caching can bypass sso
Can you please elaborate this?

Reply:
> Do you mean the crowd2-plugin caches tokens somewhere else other than in the session?

https://github.com/jenkinsci/crowd2-plugin/blob/8f647b86684d89977fd4aeaf4d0f4583b27a1eb2/src/main/java/de/theit/jenkins/crowd/CrowdConfigurationService.java#L149
I think user should be removed from this cache in case of session termination.


> Can you please elaborate this?

This will do the job
https://www.youtube.com/watch?v=LMgpuVKslw8